Vulkan rendering backend. Early Work-In-Progress

hrydgard opened the pull request:
Please note - this is not for general use, it's only an early preview. It's very rough, only works in non-buffered (crashes in buffered) and is missing many features and optimizations, so it's not even fast yet.

EDIT: At this point, it already beats OpenGL in nearly everything I've tried, though is not feature complete of course.

But, it's a start, and this will hopefully let us take performance to the next level on Android, eventually - once drivers supporting Vulkan are released, and present on devices. Some devices will get Vulkan with a Marshmallow update (nVidia Shield/TV) while others will have to wait longer or won't get it at all.

hrydgard committed:
Initial vulkan code.

This was squashed from nine commits but using old versions of Vulkan.

hrydgard committed:
More vulkan fixing

hrydgard committed:
Fixes in VulkanContext

hrydgard committed:
Enable vulkan validation

hrydgard committed:
More vulkan in thin3d

hrydgard committed:
More fixes

hrydgard committed:
Just some work on the shader generators

hrydgard committed:
Further vulkan stuff

hrydgard committed:
More progress

hrydgard committed:
Update VulkanContext

hrydgard committed:
Support multiple texture formats

hrydgard committed:
Per-frame descriptor pools. Safe-delete more types.

hrydgard committed:
Show Vulkan information in system information.

hrydgard committed:
Back to work on the PSP renderer

hrydgard committed:
Start porting TextureCache. Lots of stubbing going on.

hrydgard committed:
Fill out GPU_Vulkan.cpp with what's mostly a copy of GLES_GPU.cpp

hrydgard committed:
Reaches the first clear

hrydgard committed:
Okay, reached the shader generators

hrydgard committed:
First PSP shaders actually compile

hrydgard committed:
Implement a trivial SPIR-V disassembler, just for fun

hrydgard committed:
Some descriptor binding fixes

hrydgard committed:
Fix clearing bug, add FIFO_RELAXED flag, etc

hrydgard committed:
Begin/End frame fixes

hrydgard committed:
Pipelines, samplers, description sets, oh my

hrydgard committed:
Get untextured drawing working! (at least cube.elf)

hrydgard committed:
Separate uniform updates from shader updates

hrydgard committed:
Fix indexed drawing, flip rendering the right side up, enable culling

hrydgard committed:
Vertex format fixes. Flip through drawing the right way up.

hrydgard committed:
Fix indexed through-mode drawing

hrydgard committed:
Remove remains of FragmentTestCache from Vulkan backend

hrydgard committed:
Fix primitive topology. Crashfix.

hrydgard committed:
Work on texture support, cube.elf works correctly. Delete unused code.

hrydgard committed:
More texture work

hrydgard committed:
Fix issues with texture replacement

hrydgard committed:
Switch to "Mailbox" frame submission if available, lets us go beyond 60hz for unthrottle

hrydgard committed:
Assorted fixes and cleanups.

hrydgard committed:
Plug the texture memory leak

hrydgard committed:
Fix mixup with fragment shader ubo variables

hrydgard committed:
Add Vulkan to Win32 menu. disable gfx logging in debug by default.

hrydgard committed:
More zero-initialization.  Fix scope issue, MSVC's optimizer was aggressive enough that this was a problem. Fixes textures in release mode.

hrydgard committed:
The settings to the vertex decoder must match the implementation..

Fixes some character scale issues.

hrydgard committed:
More clearing fixes. Also fix some memory leaks.

hrydgard committed:
Delete pipeline caches asynchronously too

hrydgard committed:
Crashfix, fix left-behind culling disabling

hrydgard committed:
WIP trying to fix drawing issues

hrydgard committed:
Expose another couple of vulkan device features.

hrydgard committed:
WIP: Start work on getting the Vulkan code to compile for Android

hrydgard committed:
Android: Vulkan builds but doesn't link as we have no link library.

hrydgard committed:
Vulkan: Don't try to overlap proj with proj_through, will need a different approach.

Also, assorted bugfixes.

hrydgard committed:
Cleanups. Don't loop over bones in vshader, causes trouble

hrydgard committed:
Cleanup

Anuskuss commented:
Awesome :+1:

GAMELASTER commented:
D3D11 backend WIP too?

hrydgard commented:
@GAMELASTER  DX11? only vague plans, nothing concrete. Just made space in the enum, heh. Might remove it from the list.

VIRGINKLM commented:
Just a small note. Samsung Galaxy A6 (2016) supports Vulkan straight out of the box and NVidia Shield devices began recieving Dev buils supporting Vulkan. Also I bet Galaxy S7 variants will have straight out of the box support for it.

Zetchzie commented:
So will this work on ogl backend in the future and can this hold potential for better performance on damanding games on hardware that couldnt run at full speed previously?

zminhquanz commented:
https://en.wikipedia.org/wiki/Vulkan_(API)
Here is the list for the GPU from NVIDIA , AMD and PowerVR , Mali , Adreno , Intel to support Vulkan API
Device can be update to Android 6.0 to support it

unknownbrackets commented on Common/GraphicsContext.h:
```diff
@@ -22,6 +22,9 @@ class GraphicsContext {
 	virtual void Resize() = 0;
 
+	// Needs casting to the appropriate type, unfortunately. Should find a better solution..
+	virtual void *GetAPIContext() { return nullptr; }
```
Seems like this is no longer used?

-[Unknown]

unknownbrackets commented on GPU/Directx9/TextureCacheDX9.cpp:
```diff
@@ -1639,6 +1639,7 @@ TextureCacheDX9::TexCacheEntry::Status TextureCacheDX9::CheckAlpha(const u32 *pi
 	return (TexCacheEntry::Status)res;
 }
 
+// TODO: xoffset and yoffset are unused - bug?
```
Hmm, they're fixed at 0 also.  Maybe the plan was to support updating a subarea?  Sounds hard, maybe we should just remove.

-[Unknown]

unknownbrackets commented on GPU/Vulkan/DrawEngineVulkan.cpp:
```diff
@@ -0,0 +1,733 @@
+	indexGen.Setup(decIndex);
+
+	// All resources we need for PSP drawing. Usually only bindings 0 and 2-4 are populated.
+	VkDescriptorSetLayoutBinding bindings[5];
```
Should we zero initialize in case 1.1 adds more members?

-[Unknown]

hrydgard replied:
Vulkan is not extended that way for minor versions, instead, additional structs are chained using the pNext member, which is more ABI safe. Changing structs requires a major version upgrade. There's a whole chapter on how Vulkan will be extended in vkspec.pdf :)

But yeah, might as well zero it anyway.

unknownbrackets commented on Windows/MainWindow.cpp:
```diff
@@ -701,13 +702,22 @@ namespace MainWindow
+		case WM_EXITSIZEMOVE:
+			inResizeMove = false;
+			HandleSizeChange(SIZE_RESTORED);
+			break;
```
Are these changes Vulkan related or separate?

-[Unknown]

hrydgard replied:
Commited separately to master.

hrydgard committed:
Load Vulkan dynamically. Fix Android build (though - Vulkan is not working on Android yet)

hrydgard committed:
Remove remains of xcb support. Will add back later.

hrydgard committed:
Possible buildfix

hrydgard committed:
Hook up Vulkan on Android, no idea if it works. Move VulkanContext to Common.

hrydgard committed:
Turn off Vulkan in CMake-based and Qt-based builds, for now

phire committed:
VulkanLoader: Load vkDestroySurfaceKHR

phire committed:
VulkanLoader: Detect loader failure on windows.

hrydgard committed:
Show backend chooser on Android so we can choose Vulkan.

hrydgard committed:
Gets the Vulkan backend running, but not drawing correctly, on Android.

Not sure what's wrong, I get a yellow flashing screen.

unknownbrackets commented:
Where is the graphicsContext supposed to be populated?  I had to add `PSP_CoreParameter().graphicsContext = graphicsContext;` to `NativeRender()` to fix some crashes.

Well, now I'm crashing in vkCmdClearAttachments, but it's progress.

-[Unknown]

hrydgard commented:
@unknownbrackets Remember to turn off buffered rendering, it will crash :)

Don't recall having problems with graphicsContext... hm

hrydgard committed:
Re-fix depth buffer initialization

hrydgard commented:
@unknownbrackets  turns out I broke depth buffer init, too, when messing with the shield

hrydgard committed:
Buildfix (glslang project settings /MT)

hrydgard committed:
Vulkan: We're doing dual src blending wrong (GLSL part, not sure how to do it right) so disable it.

hrydgard committed:
Vulkan: Attempt at fixing dual source blending. Should work but doesn't.

unknownbrackets commented on Windows/PPSSPP.sln:
```diff
@@ -75,18 +76,30 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "glslang", "..\ext\glslang.v
 		Debug|Win32 = Debug|Win32
 		Debug|x64 = Debug|x64
+		Release CRT DLL|Win32 = Release CRT DLL|Win32
+		Release CRT DLL|x64 = Release CRT DLL|x64
```
Do we want to add these, or were they just added by something else?

-[Unknown]

hrydgard replied:
Hm, they were probably added when I tried to import the CMake-generated project for glslang, before I made my own instead.

unknownbrackets commented on Windows/WindowsHost.cpp:
```diff
@@ -181,6 +187,7 @@ void WindowsHost::SetDebugMode(bool mode) {
 }
 
 void WindowsHost::PollControllers(InputState &input_state) {
+	return;
```
Temp hack warning.

-[Unknown]

hrydgard replied:
oops indeed

hrydgard committed:
Remove redundant configuration, remove temp hack

unknownbrackets commented on unittest/UnitTests.vcxproj:
```diff
@@ -123,7 +122,7 @@
       <SubSystem>Console</SubSystem>
       <GenerateDebugInformation>true</GenerateDebugInformation>
-      <AdditionalDependencies>Ws2_32.lib;winmm.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;opengl32.lib;dsound.lib;glu32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;comctl32.lib;d3d9.lib;dxguid.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>..\Windows\x64\Debug\glslang.lib;..\ext\vulkan\vulkan-1.lib;Ws2_32.lib;winmm.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;opengl32.lib;dsound.lib;glu32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;comctl32.lib;d3d9.lib;dxguid.lib;%(AdditionalDepen ...
```
Any reason only this one gets the glslang lib?

-[Unknown]

hrydgard replied:
It's not supposed to be there, it's linked through "References". this was an old thing that slipped through some cleanup.

unknownbrackets committed:
Move Vulkan code into a filter.

unknownbrackets committed:
Don't crash as hard on shader compile error.

unknownbrackets committed:
Fix GE debugger preview in non-buffered.

unknownbrackets commented:
@hrydgard
> Refactor away nextTexture_. Not needed on Vulkan.

Do you mean because you plan to remove the hardware transform -> software fallback and move texture decoding earlier so that it can be done before `SetTexture()`?  Because that is the reason `nextTexture_` exists.  We don't really need it for D3D9 either, technically, I suppose.

To clarify, `ApplyTexture()` MUST be called AFTER the vertices have been decoded.  I'm not sure, but right now this isn't happening, and this could be contributing to rendering bugs.  At the very least, it's ignoring the output of the vertex decode (texture bounds.)

-[Unknown]

hrydgard commented:
Yeah, I intend to move things around a bit, I've just been swamped. And yes, UV scale/offset seems a bit broken, among other things, as expected...

unknownbrackets committed:
Use draws for alpha/stencil only clear in Vulkan.

unknownbrackets committed:
Specify a useful app version when loading Vulkan.

hrydgard committed:
Should not specify the Vulkan patch version when requesting Vulkan 1.0.

hrydgard committed:
Update Vulkan headers to 1.0.5. Remove references to vulkan-1.lib as we load dynamically.

hrydgard committed:
Fix a bunch of new vulkan validation layer failures.

hrydgard committed:
Work around what seems like a validation layer bug (trigger by scrolling
around the UI a bit)

hrydgard committed:
Turn off annoying "debug flicker" I used to see if we were swapping buffers properly early on

Also minor logging changes, refine the false-positive debug layer check

hrydgard committed:
Replace ugly workaround with VK_LAYER_LUNARG_unique_objects which prevents handle reuse.

See https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/121

unknownbrackets committed:
Handle Vulkan init errors more gracefully.

hrydgard committed:
Typo fix

hrydgard committed:
Revert "Replace ugly workaround with VK_LAYER_LUNARG_unique_objects which prevents handle reuse."

This reverts commit 6342c07a19d5ab410cbc233a570488214923a9f8.

unknownbrackets commented on Core/Config.h:
```diff
@@ -50,10 +50,14 @@ enum BufferFilter {
 	OPENGL = 0,
 	DIRECT3D9 = 1,
+  DIRECT3D11 = 2,
+  VULKAN = 3,
```
Indent nitpick.

-[Unknown]

unknownbrackets committed:
Allow the graphicsContext to be overridden.

This is useful when coming from, for example, headless.

unknownbrackets committed:
Switch headless to using graphics contexts.

Vulkan doesn't work yet, needs some thin3d for the command buffer?

unknownbrackets commented:
Personally, I would prefer this convention:

```c++
  VkFramebufferCreateInfo fb_info = { VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO };
  fb_info.renderPass = surface_render_pass_;
```

Instead of:
```c++
  VkFramebufferCreateInfo fb_info = {};
  fb_info.sType = VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO;
  fb_info.pNext = NULL;
  fb_info.renderPass = surface_render_pass_;
```

What do you think?

-[Unknown]

hrydgard commented:
Hm, not bad. I guess the compiler will eliminate the extra stores (first zeroing the struct, then filling it out) anyway. Let's do that.

hrydgard committed:
Fix UV scale for some vertex formats (like GL). Separate updates of PROJ and PROJTHROUGH matrices.

hrydgard committed:
Silence another validation warning

hrydgard committed:
Mem leak fixes, cleanups

hrydgard committed:
Fix bug where we sometimes pushed too much vertex data.

Doesn't seem to fix much though...

unknownbrackets committed:
Disable last texture reuse for now.

This will be slower, but otherwise a lot of textures are missing in
various games, like Final Fantasy 4.  Better to get all the bugs out first
and then start optimizing.

Note: there is still missing drawing in Final Fantasy 4, but only text
after this change.

hrydgard committed:
Enable the Unpack Subimage path on DX9 (maybe should just remove the check)

hrydgard committed:
Matrix depth fix. Still have depth issues though.

hrydgard committed:
Default validation on in Debug and off in Release

hrydgard committed:
Add a secondary way to upload textures - through buffers.

hrydgard committed:
Upload PSP textures through a push buffer instead of image copy. More flexible.

hrydgard committed:
Vulkan: Enable mipmapping

hrydgard committed:
Show how much pushbuffer space is used each frame.

Optimizing this number (by avoiding redundant UBO uploads etc) will probably help performance.

unknownbrackets commented:
Do you think it would be a good idea to decode directly onto the push buffer (from the jit), and then if we're caching, issue a cmd to copy that to the cache buffer (which would then be used for subsequent draws?)

By the way, I'm seeing something funny with vertex coords in FF4, seems to be why only the first letter is drawing in some places.  Still debugging, but this area has me thinking about vertex decode is all.

-[Unknown]

unknownbrackets commented on GPU/Vulkan/ShaderManagerVulkan.cpp:
```diff
@@ -162,7 +162,7 @@ static void ConvertProjMatrixToVulkan(Matrix4x4 &in, bool invertedX, bool invert
 		yoff = -yoff;
 
-	const Vec3 trans(xoff, yoff, gstate_c.vpZOffset + 0.5f);
+	const Vec3 trans(xoff, yoff, gstate_c.vpZOffset * 0.5f + 0.5f);
```
Wait wait wait, does Vulkan use the half-pixel offset insanity that Direct3D9 does?  It doesn't right?

-[Unknown]

-[Unknown]

unknownbrackets commented on GPU/Vulkan/ShaderManagerVulkan.cpp:
```diff
@@ -0,0 +1,581 @@
+	if (invertedX)
+		xoff = -xoff;
+	if (invertedY)
+		yoff = -yoff;
```
I don't remember properly why we do 5667f348055c59eaee51a2498d0a0b1c3de3be26, but I don't think we need/want to do it in Vulkan?  This is what causes FF4 to not render some stuff.

-[Unknown]

hrydgard replied:
Vulkan does not need the half-pixel offset, should remove that indeed.

Vulkan supports very large viewports on all hardware I've seen limits for on gpuinfo.org so we should be able to get rid of the viewport clamping stuff, and in that case the offset should be 0 and this shouldn't matter.

Are you saying FF4 renders fully if you remove these lines?

unknownbrackets replied:
Yes, I didn't find anything else missing with those lines removed.  They are causing it to render some text off screen.  We don't do this in GLES.

Also, the offset will always be the drift from left to right, so it ought to always be positive, no?  I'm trying to understand why this would be correct on Direct3D 9.  Even if the coordinate system is negated, if the center is toward the right, the final coordinates should be translated positive...

-[Unknown]

hrydgard replied:
Can't figure it out either, I think it was only meant to negate the half-pixel offset, not the clipping offset...

unknownbrackets replied:
Hmm.  Now that I think harder about it, should the half-pixel offset be negated either?  It is also a drift in the final mapping of the pixel to screen coordinates.  It ought to have the same properties...

Also, it's being negated twice in vulkan/d3d9...

-[Unknown]

hrydgard committed:
Get rid of strange offset in Vulkan matrix converter

unknownbrackets committed:
Vulkan: Make sure depth==stencil when clearing.

We have to clear both at the same time.  I think it makes sense to
consider this part of the "separate alpha" flag, since alpha has to match
both color and depth.

unknownbrackets committed:
Vulkan: Properly set the alpha test ref.

It's definitely not a float, oops.

unknownbrackets committed:
Vulkan: Simplify color testing.

unknownbrackets committed:
Correct the 5551 format for now.

Otherwise we get wrong colors - also we are still converting colors.

hrydgard committed:
Split out VulkanTexture from VulkanContext.cpp/h into VulkanImage.cpp/h

unknownbrackets committed:
Cleanup differences in aniso handling.

Some places were not handling it as a power of two.

unknownbrackets committed:
Vulkan: Initial support for aniso filtering.

hrydgard committed:
Vulkan: Fix our use of dual source blending, re-enable if available.

unknownbrackets commented:
It's rendering now on my Android TV, although rotated 180 degrees.  I'm not sure what fixed it, do you know?

-[Unknown]

hrydgard commented:
Yeah. Flipped vertically sounds bad though...

hrydgard committed:
Use separate pushbuffers for UBO/Index/Vertex. This will make decoding directly into the pushbuffers easier, plus other benefits later.

(For example, we'll often be able to avoid rebinding the vertex and
index buffers at new offsets by just keep counting upwards between draws,
if the vertex format is the same but other state changed)

hrydgard commented:
Ah, the upside down thing (an upside down mirror, not 180 degrees) is just that part of the emulator thinks it's in Vulkan when it's not, so it doesn't flip the projection matrix.

unknownbrackets commented:
Yeah, sorry, I corrected my comment right after I made it - at first I assumed it was 180 degrees because of some input problems, then when I looked again I realized it was just mirrored (I'm not sure about the input problem but I can no longer reproduce, maybe I was just tired... it wasn't accepting most arrows and was doing odd things.)

That's also when I realized it must just be using GL just as you say, and edited my comment for that as well.

-[Unknown]

hrydgard committed:
Move VulkanPushBuffer into VulkanMemory.h

hrydgard committed:
Change the VulkanPushBuffer API to allow for adding support for dynamic growth

hrydgard committed:
Always specify depthstencil state in pipelines.

Turns out it's only OK to leave out if rendering only to color,
otherwise behaviour is undefined - which means it may still be enabled.

hrydgard committed:
Fix a resource leak. Request COHERENT memory for pushbuffers.

hrydgard committed:
Minor optimizations

hrydgard committed:
No need to align vertex/index data. Will only be harmful for an upcoming optimization.

hrydgard committed:
Vulkan: Remove support for software skinning to reduce complexity

hrydgard committed:
Decode vertex data directly into the vertex pushbuffer, saving a memcpy.

hrydgard committed:
Be more economical with UBO pushbuffer space by reusing the last data when possible.

hrydgard committed:
SoftGPU with Vulkan runs but displays black

hrydgard committed:
Vulkan: Add ugly temporary warning about buffered rendering not working

hrydgard committed:
Combined two uniforms to get the base UBO down to 512b, in order to not waste space (nVidia needs 256-byte alignment)

hrydgard committed:
Buildfix to last commit, add a comment
